[webkitscmpy] Only respect the latest review

JonWBedard opened the pull request:
#### 3bced9d602295f75e2c637f4ccc6d1e733a6af0a
<pre>
[webkitscmpy] Only respect the latest review
<a href="https://bugs.webkit.org/show_bug.cgi?id=231987">https://bugs.webkit.org/show_bug.cgi?id=231987</a>
&lt;rdar://problem/84434991 &gt;

Reviewed by Dewei Zhu.

GitHub will report the entire history of a review, we should only
respect the latest review state.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
(GitHub.PRGenerator.reviewers):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
(TestNetworkPullRequestGitHub.test_approved_edits):
</pre>

dpogue commented:
WebKit isn't developed on GitHub (this is a read-only mirror), so the WebKit developers are unlikely to see this.

Please file a bug against WebKit at https://bugs.webkit.org/ and against iOS at https://bugreport.apple.com/

rniwa commented:
Please report this problem to Apple instead: https://developer.apple.com/bug-reporting/

rniwa commented:
Closing.

kenygamer commented:
Took a while to answer...
Was the issue fixed anyways? It didn't occur me anymore since some iOS update, that is, I still present browsing issues in certain webpages sometimes, but that seems to be a normal occurrence. What is worrying is that if it happens again it can really cause a bad moment to the person who has it, and it did to me. I had to go to technical service and they (forcefully) restored the device, their response was a "software issue".

rniwa commented:
We had a number of issues in service worker and cache that had since addressed. Since there is no detailed reproducible test case attached, I can't tell you which one of the existing bugs you were encountering but such a bug report & discussion really needs to happen on https://bugs.webkit.org and/or https://feedbackassistant.apple.com/welcome, not on a Github PR.

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
```diff
@@ -172,13 +172,17 @@ def reviewers(self, pull_request):
+            state_for = {}
             for review in self.repository.request('pulls/{}/reviews'.format(pull_request.number)):
-                contributor = self._contributor(review['user']['login'])
+                state_for[self._contributor(review['user']['login'])] = review.get('state')
```
We assume reviews from `self.repository.request('pulls/{}/reviews'.format(pull_request.number))` are ordered by time from oldest to newest, right?

JonWBedard replied:
Yes!

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
```diff
@@ -172,13 +172,17 @@ def reviewers(self, pull_request):
+                if status == 'APPROVED' and (not needs_status or contributor.status == Contributor.REVIEWER):
                     pull_request._approvers.append(contributor)
-                elif review.get('state') == 'CHANGES_REQUESTED':
+                elif status == 'CHANGES_REQUESTED':
```
This whole change seems worth a unit test.

JonWBedard replied:
We are already partially covered by unit tests, but I added a more explicit one.

dewei-zhu replied:
Great.

JonWBedard committed:
[webkitscmpy] Only respect the latest review
https://bugs.webkit.org/show_bug.cgi?id=231987
<rdar://problem/84434991>

Reviewed by Dewei Zhu.

GitHub will report the entire history of a review, we should only
respect the latest review state.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
(GitHub.PRGenerator.reviewers):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
(TestNetworkPullRequestGitHub.test_approved_edits):
